File: wupc2020/g.cpp
#include <bits/stdc++.h>
using namespace std;

struct dat {
  long long t, p;
};

int n, k, t;
vector<dat> v;
vector<long long> dp;

long long solve();

int main() {
  cin >> n >> k >> t;
  v.resize(n);
  for (auto& p : v) cin >> p.t >> p.p;
  cout << solve() << endl;
  return 0;
}

long long solve() {
  dp.assign(t + 1, 0);
  for (int i = 0; i < n; ++i)
    for (int j = t; j >= v[i].t; --j)
      dp[j] = max(dp[j], dp[j - v[i].t] + v[i].p);
  v.resize(t + 1);
  for (int i = 0; i <= t; ++i) v[i].t = i + k, v[i].p = dp[i];
  for (int i = 0; i <= t; ++i)
    for (int j = v[i].t; j <= t; ++j)
      dp[j] = max(dp[j], dp[j - v[i].t] + v[i].p);
  long long res = 0;
  for (auto p : dp) res = max(res, p);
  return res;
}

File: wupc2020/g.py
N, K, T = map(int,input().split())
t = [0] * N
p = [0] * max(N,T + 1)
for i in range(N):
    t[i], p[i] = map(int,input().split())
dp = [0] * (T + 1)
for i in range(N):
    for j in range(T, t[i] - 1 , -1):
        dp[j] = max(dp[j],dp[j-t[i]] + p[i])
for i in range(T + 1):
    p[i] = dp[i]
for i in range(T + 1):
    nt = i + K
    for j in range(nt, T + 1):
        dp[j] = max(dp[j],dp[j-nt] + p[i])
res = 0
for i in range(T + 1):
    res = max(res,dp[i])
print(res)